### PR TITLE
feat(source): add support for creating stripe Sources

### PR DESCRIPTION
**Pull request by MohammedBashiru**
## What is the current behavior?

There is no support for creating stripe **Sources**. 
reference https://stripe.com/docs/mobile/ios/sources

## What is the new behavior?

Added support for creating stripe **Sources** for iOS and Android
reference 
https://stripe.com/docs/mobile/ios/sources
https://stripe.com/docs/mobile/android/sources

Fixes/Implements/Closes #[Issue Number].
https://github.com/triniwiz/nativescript-stripe/issues/105

### CODE REVIEW

File: src/index.d.ts
Comment (RobertGardner) [inline src/index.d.ts]: Instead of specifying 'null | string', for consistency with the rest of the code, I suggest using '?', like this:

city?: string;

File: src/index.d.ts
Comment (RobertGardner) [inline src/index.d.ts]: Naming convention is to use camelcase rather than underscores. So postalCode instead of postal_code. Do this throughout.

File: src/index.d.ts
Comment (RobertGardner) [inline src/index.d.ts]: You can use 'object' instead of '{}'

File: src/index.d.ts
Comment (RobertGardner) [inline src/index.d.ts]: I'm not sure what this is since it doesn't exist in either the Android or iOS fields. My investigation suggested this field is not useful?

File: src/index.d.ts
Comment (RobertGardner) [inline src/index.d.ts]: To be consistent with Token and other APIs, this should be Date instead of number.

File: src/index.d.ts
Comment (RobertGardner) [inline src/index.d.ts]: I believe this should be CardAddress instead of string.

File: src/stripe.android.ts
Comment (RobertGardner) [inline src/stripe.android.ts]: This isn't mapped properly. It needs to be `source.getAmount().longValue()`.

The easiest way to find these mapping errors is to add a type to the `source` parameter. Like this:
`function (source: source: com.stripe.android.model.Source) =>`

(This should have been done on the Token function above. If you add it, I'd appreciate it.)

File: src/stripe.android.ts
Comment (RobertGardner) [inline src/stripe.android.ts]: For some reason `isLiveMode()` is spec'd to be a Java Boolean instead of boolean, so it needs to be mapped as `source.isLiveMode().booleanValue()`. It wouldn't hurt to verify in the debugger if the typings are correct.

File: src/stripe.android.ts
Comment (RobertGardner) [inline src/stripe.android.ts]: After you change `created` to be a Date, the mapping will be `new Date(source.getCreated().toString())` (see how it's done in Token above).

File: src/stripe.android.ts
Comment (RobertGardner) [inline src/stripe.android.ts]: This needs to be explicitly mapped. It's a pain. :-)  Is `owner` a field you need on the client?

File: src/stripe.ios.ts
Comment (RobertGardner) [inline src/stripe.ios.ts]: Use `source: STPSource` to find all the mapping problems.

File: src/stripe.ios.ts
Comment (RobertGardner) [inline src/stripe.ios.ts]: Enums on iOS are a real pain. They come through as integers, which need to be mapped to the proper string values.

**Review by RobertGardner — CHANGES_REQUESTED** [review]
For consistency with the existing style, I've made some style comments. Also, lint is complaining about missing semicolons and spaces at the ends of some of the lines.

We need to expose this in the Angular demo, as well. I can do that for you once everything is ready if you'd like.

It looks like there are some mapping bugs in the Android implementation and quite a few in the iOS implementation. See comments below.

In general, since the mapping can be tedious, you might want to limit the fields you expose in the plug-in to just the ones that you need or are likely to need.

Finally, once we're ready to submit, the package version will need to be incremented (probably to 6.4.0 since new functionality was added) and the CHANGELOG updated.

File: src/index.d.ts
Comment (MohammedBashiru) [reply src/index.d.ts]: resolved

File: src/stripe.ios.ts
Comment (MohammedBashiru) [reply src/stripe.ios.ts]: Sorry I do not understand this.

**Comment by MohammedBashiru** [comment]
> We need to expose this in the Angular demo, as well. I can do that for you once everything is ready if you'd like.

Yes you can do that. Am not familiar with Angular

> Finally, once we're ready to submit, the package version will need to be incremented (probably to 6.4.0 since new functionality was added) and the CHANGELOG updated.

Sure I'd leave that to you as well.

File: src/index.d.ts
Comment (RobertGardner) [inline src/index.d.ts]: Remove `status`?

File: src/stripe.android.ts
Comment (RobertGardner) [inline src/stripe.android.ts]: Imports get ordered alphabetically by the `from` source, and the names within `{}` are also alphabetized.

What that means here is move `import *`... to line 2 (this one may have been wrong in the original source), and put `Source` before `StripePaymentIntentCommon`.

File: src/stripe.android.ts
Comment (RobertGardner) [inline src/stripe.android.ts]: Add a trailing semicolon.

File: src/stripe.android.ts
Comment (RobertGardner) [inline src/stripe.android.ts]: Need space after `function` and before `(`.

File: src/stripe.android.ts
Comment (RobertGardner) [inline src/stripe.android.ts]: Space before '(' and after ')'.

File: src/stripe.common.ts
Comment (RobertGardner) [inline src/stripe.common.ts]: These 3 fields (`status`, `type`, `usage`) are never set. Remove them?

File: src/stripe.common.ts
Comment (RobertGardner) [inline src/stripe.common.ts]: Minor, but I find it useful to add a comment to the end of this line
`// in pennies`

The same should be added in `index.d.ts` to keep them consistent. I actually think most people using the plugin will look first in `index.d.ts` so that one is more important.

File: src/stripe.ios.ts
Comment (RobertGardner) [inline src/stripe.ios.ts]: Likewise here, put `Source` before `StripePaymentIntentCommon`.

File: src/stripe.ios.ts
Comment (RobertGardner) [inline src/stripe.ios.ts]: Trailing semicolon.

**Review by RobertGardner — CHANGES_REQUESTED** [review]
Looks good. Thanks!

A few minor comments.

Lint errors are noted below. Our continuous integration runs Lint, so those need to be fixed before merging.

There are some fields in Source that aren't set in either the Android or iOS implementations, so they probably should be removed from the Source interface.

The Source interface in index.d.ts doesn't match that in the implementations.

**Comment by triniwiz** [comment]
Atm it looks like you are only creating a card source also there are other source types you are allowed to create It would be nice to have all options available 
IOS : - https://stripe.dev/stripe-ios/docs/Classes/STPSourceParams.html
Android :- https://stripe.dev/stripe-android/com/stripe/android/model/SourceParams.html

**Comment by RobertGardner** [comment]
@triniwiz Yes, we should support all the source types, especially Google Pay and Apple Pay. I'd like to get this submitted first with card support, and then add support for the other types in subsequent PRs. I can do some of that work, as well.

**Review by RobertGardner — APPROVED** [review]
I'm not sure why you don't see those lint errors in VS Code. I see "trailing white space" in both `stripe.ios.ts` and `stripe.android.ts`. That's a very annoying lint error. Maybe I should turn it off!

Anyway, since I have to update the package version number and CHANGELOG anyway, I'll go ahead and merge this PR and remove those lint errors in that commit. It's not worth another back-and-forth just for trailing whitespace!!

By the way, you can see that these trivial lint errors are causing the build to fail by examining our Travis continuous integration build at https://travis-ci.org/github/triniwiz/nativescript-stripe.

Once again, thanks for doing this work.

**Pull request merged**
